docs/hub/webhooks-guide-auto-retrain.md

```diff
@@ -0,0 +1,176 @@
+
+This Space uses Docker, Python, [FastAPI](https://fastapi.tiangolo.com/) and [uvicorn](https://www.uvicorn.org) to run a simple HTTP server. Read more about Docker Spaces [here](https://huggingface.co/docs/hub/spaces-sdks-docker).
+
+
+The entry point is [src/main.py](https://huggingface.co/spaces/huggingface-projects/auto-retrain/blob/main/src/main.py). Let's walk through this file and detail what it does:
+
+1. It spawns a FastAPI app that will listen to HTTP `POST` requests on `/webhook`:
+
+```python
+from fastapi import FastAPI
+
+# [...]
+@app.post("/webhook)
+async def post_webhook(
+	# ...
+):
+	# ...
+```
+
+2. This route checks that the `X-Webhook-Secret` header is present, and that its value is the same as the one you set in your webhook's settings. The `WEBHOOK_SECRET` secret must be set in the Space's settings, and be the same as the secret set in your webhook.
+
+```python
+# [...]
+
+WEBHOOK_SECRET = os.getenv("WEBHOOK_SECRET")
+
+# [...]
+
+@app.post("/webhook")
+async def post_webhook(
+	# [...]
+	x_webhook_secret:  Optional[str] = Header(default=None),
+	# ^ checks for the X-Webhook-Secret HTTP header
+):
+	if x_webhook_secret is None:
+		raise HTTPException(401)
+	if x_webhook_secret != WEBHOOK_SECRET:
+		raise HTTPException(403)
+	# [...]
+```
+
+3. The event's payload is encoded as JSON. Here, we'll be using pydantic models to parse the event payload. We also specify that we will run our webhook only when:
+- the event concerns the input dataset
+- the event is an update on the repo's content, i.e. there has been a new commit
+
+
+```python
+# defined in src/models.py
+class WebhookPayloadEvent(BaseModel):
+	action: Literal["create", "update", "delete"]
+	scope: str
+
+class WebhookPayloadRepo(BaseModel):
+	type: Literal["dataset", "model", "space"]
+	name: str
+	id: str
+	private: bool
+	headSha: str
+
+class WebhookPayload(BaseModel):
+	event: WebhookPayloadEvent
+	repo: WebhookPayloadRepo
+
+# [...]
+
+@app.post("/webhook)
+async def post_webhook(
+	# [...]
+	payload: WebhookPayload,
+	# ^ Pydantic model defining the payload format
+):
+	# [...]
+	if not (
+		payload.event.action == "update"
+		and payload.event.scope.startswith("repo.content")
+		and payload.repo.name == config.input_dataset
+		and payload.repo.type == "dataset"
+	):
+		# no-op if the payload does not match our expectations
+		return {"processed": False}
+	#[...]
+```
+
+4. If the payload is valid, the next step is to create a project on AutoTrain, schedule a fine tuning of the input model (`microsoft/resnet-50` in our example) on the input dataset, and create a discussion on the dataset when it's done!
+
+```python
+def schedule_retrain(payload: WebhookPayload):
+	# Create the autotrain project
+	try:
+		project = AutoTrain.create_project(payload)
+		AutoTrain.add_data(project_id=project["id"])
+		AutoTrain.start_processing(project_id=project["id"])
```

Comment:
you don't go over how to import or define `class AutoTrain ` to interact with AutoTrain through its API but i guess it's on purpose? as it's not the focus of this guide?

Reply:
(maybe at some point we should revive the SDK to expose a super simple wrapper. Or even just put it inside `huggingface_hub`... cc @Wauplin @abhishekkrthakur ) or just link to a Gist with the code..

Reply:
> you don't go over how to import or define class AutoTrain  to interact with AutoTrain through its API but i guess it's on purpose? as it's not the focus of this guide?

Correct!

Reply:
(there's a lot of code in https://github.com/huggingface/autotrain-advanced public repo but i'm not sure everything is still needed or up to date)